framework/include/utils/InputParameters.h

```diff
@@ -1015,6 +1015,7 @@ class InputParameters : public Parameters
   friend InputParameters emptyInputParameters();
   friend class InputParameterWarehouse;
   friend class Parser;
+  friend class ActionWarehouse;
```

Comment:
and this?

Reply:
This one is for that printInputFile function.

Reply:
```suggestion
  friend void ActionWarehouse::printInputFile(std::ostream & out);
```
Let's make it specific to where we need it then

Reply:
> Unfortunately we cannot include ActionWarehouse.h in InputParameters.h because of the cyclic including I think. Any suggestions?

Now that you are storing a reference to the input parameters in `Action`, you should be able to change the `#include "InputParameters.h` in that file to a forward declaration which would allow you to include the action warehouse header in `InputParameters.h`

Reply:
Let me try. But I've found there are bunch of other places indirectly included by Action.h having includes of InputParameters.h.

Reply:
Unfortunately, there are some template functions in `Action.h` relying on `InputParameters.h`... Same for the `Registry.h`, included by `Action.h`. I remove the unnecessary include of `Registry.h` in `Action.h` and few other indirect includes, but we cannot remove the include in `Action.h` itself. Also we cannot remove the include of `Action.h` in `ActionWarehouse.h`. So I think the possible solution could be move this function out of `ActionWarehouse`, for example in the Parser. But that requires the action warehouse to provide a public function to expose all its actions, which may not be a good idea either. Another solution is to let the `InputParameterWarehouse` to do the print, but that could be other implications. I also feel maybe we should just let parser do the print.

So I suggest we keep this. Are you convinced? I will push what I have on some attempts of reducing header includes in the last commit.

Reply:
fine